src/merkle/tiered_smt/mod.rs

```diff
@@ -0,0 +1,425 @@
+    pub fn with_leaves<I, T>(self, leaves: I) -> Result<Self, StorageError>
+    where
+        I: IntoIterator<Item = (Word, Word)>,
+        Leaf: From<T>,
+    {
+        leaves.into_iter().try_fold(self, |mut tree, (key, value)| {
+            tree.insert(key, value)?;
+            Ok(tree)
+        })
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns a leaf value indexed by the provided key.
+    pub fn get_leaf_value<K>(&self, key: K) -> Result<Option<Word>, StorageError>
+    where
+        CanonicalWord: From<K>,
+    {
+        let key = CanonicalWord::from(key);
+        self.storage.get_leaf_value(&key)
+    }
+
+    /// Returns the current Merkle root of the state of the tree.
+    pub const fn root(&self) -> &Word {
+        &self.root
+    }
+
+    // PROVIDERS
+    // --------------------------------------------------------------------------------------------
+
+    /// Fetch a tree node, returning an empty sub-tree constant if absent.
+    pub fn get_node(&self, index: &NodeIndex) -> Result<Word, StorageError> {
+        debug_assert!(index.depth() <= Self::MAX_DEPTH);
+        let node = self
+            .storage
+            .get_node(index)?
+            .or_else(|| {
+                EMPTY_SUBTREES
+                    .get(index.depth() as usize)
+                    .copied()
+                    .map(Word::from)
+            })
+            .unwrap_or_default();
+        Ok(node)
+    }
+
+    /// Builds a path from the node to the root.
+    pub fn get_leaf_path<K>(&self, key: K) -> Result<MerklePath, StorageError>
+    where
+        CanonicalWord: From<K>,
+    {
+        self.get_leaf_index(key)
+            .and_then(|node| self.get_node_path(node))
+    }
+
+    /// Builds a path from the node to the root.
+    pub fn get_node_path(&self, mut index: NodeIndex) -> Result<MerklePath, StorageError> {
+        let mut path = Vec::with_capacity(index.depth() as usize);
+        while index.depth() > 0 {
+            let sibling = self.get_node(&index.sibling())?;
+            path.push(sibling);
+            index.backtrack();
+        }
+        Ok(MerklePath::new(path))
+    }
+
+    /// Returns the index of the leaf with the specified key.
+    ///
+    /// If the leaf is not in the tree, returns the index at which it could have been located.
+    /// Specifically, this could result in indexes of the following:
+    /// - An empty node.
+    /// - Another leaf which shares a prefix with the specified key.
+    pub fn get_leaf_index<K>(&self, key: K) -> Result<NodeIndex, StorageError>
+    where
+        CanonicalWord: From<K>,
+    {
+        let key = CanonicalWord::from(key);
+        let mut leaf = LeafIndex::from(key);
+        let mut node = leaf.traverse_to_first_tier();
+        self.traverse_until_empty_or_leaf(&mut leaf, &mut node)?;
+        Ok(node)
+    }
```

Comment:
One of the possible inputs of this function is a key that doesn't exist in the storage. In that case, we need to traverse the path until an empty node, and return it (where that key would expect to be inserted). That way, this can't be a simple fetch operation as tree logic (traverse between tiers) will necessarily be involved.

I would avoid as much as possible to couple any logic of the tiered tree to the storage as it would make our lives much harder in the future. Right now, we can just remove this storage and put anything else there, without the need to refactor anything, but if we couple the logic, it will not be so simple.